csm_web/scheduler/models.py

```diff
@@ -30,150 +35,118 @@ class Attendance(models.Model):
+        if self.section:
+            return f"{self.name} ({self.section.course.name})"
+        return self.name
+
     class Meta:
-        unique_together = ("name",)
-
-
-class Profile(ActivatableModel):
-    STUDENT = "ST"
-    JUNIOR_MENTOR = "JM"
-    ASSOCIATE_MENTOR = "AM"
-    SENIOR_MENTOR = "SM"
-    COORDINATOR = "CO"
-    ROLE_CHOICES = (
-        (STUDENT, "Student"),
-        (JUNIOR_MENTOR, "Junior Mentor"),
-        (ASSOCIATE_MENTOR, "Associate Mentor"),
-        (SENIOR_MENTOR, "Senior Mentor"),
-        (COORDINATOR, "Coordinator"),
-    )
-    ROLE_MAP = dict(ROLE_CHOICES)
+        abstract = True
 
-    leader = models.ForeignKey(
-        "self",
+
+class Student(Profile):
+    section = models.ForeignKey(
+        "Section",
         on_delete=models.CASCADE,
-        related_name="followers",
         blank=True,
         null=True,
-    )
-    course = models.ForeignKey(Course, on_delete=models.CASCADE)
-    role = models.CharField(max_length=2, choices=ROLE_CHOICES, blank=True)
-    user = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE)
-    section = models.ForeignKey(
-        "Section", on_delete=models.CASCADE, blank=True, null=True
+        related_name="students",
     )
 
-    def clean(self):
-        if self.section:
-            mentor_set = self.section._get_mentor_set()
-            if mentor_set.count() > 1:
-                raise ValidationError(_("Cannot save section with more than 1 mentor!"))
-            if self.course != self.section.course:
-                raise ValidationError(_("Profile course must match section course!"))
+    class Meta:
+        unique_together = ("user", "section")
 
-    @property
-    def name(self):
-        return f"{self.user.first_name} {self.user.last_name}"
 
-    def __str__(self):
-        return f"{self.course.name} {self.role} - {self.name}"
+class Mentor(Profile):
+    pass
 
 
-class Section(models.Model):
+class Section(ValidatingModel):
     course = models.ForeignKey(Course, on_delete=models.CASCADE)
-    default_spacetime = models.OneToOneField("Spacetime", on_delete=models.CASCADE)
+    spacetime = models.OneToOneField("Spacetime", on_delete=models.CASCADE)
     capacity = models.PositiveSmallIntegerField()
-
-    def _get_mentor_set(self):
-        # excluding STUDENT doesn't work because of empty role
-        return self.profile_set.filter(
-            role__in=[
-                Profile.JUNIOR_MENTOR,
-                Profile.ASSOCIATE_MENTOR,
-                Profile.SENIOR_MENTOR,
-                Profile.COORDINATOR,
-            ]
-        )
-
-    @property
-    def students(self):
-        return self.profile_set.filter(role=Profile.STUDENT)
-
-    @property
-    def mentor(self):
-        mentor_profiles = mentor_profiles = self._get_mentor_set()
-        assert mentor_profiles.count() <= 1
-        return mentor_profiles.first()
+    mentor = models.ForeignKey(Mentor, on_delete=models.SET_NULL, blank=True, null=True)
```

Comment:
do we want mentor to be a onetoone, since that's a valid assumption as long as we don't have to deal with coteaching?

also, I'm kind of wary of having a nullable field here, where use cases looking up a student's mentor's name through their section might run into problems
instead, to represent mentorless sections (if we even want to do that) we could consider creating mentor profiles that point to a "sentinel" user

Reply:
My rational behind making it nullable was for the following situations:
- Sometimes coords create sections they know they _will_ find people to teach and allow students to enroll in them, and then assign a mentor after students have enrolled
- A mentor gets sick or leaves CSM or has a conflict or something, so mentor may be null in the interim while coords find a replacement mentor for the section

No one in exec has displayed any interest in co-teaching, so I say we put that on the back-burner for now and leave it one-to-one.

And why can't we just solve the "looking up mentor name problem" in the serializer (i.e. `name = student.section.mentor.name if student.section.mentor else 'TBD'`)

Reply:
my issue is there might be some other use case where we need to look up a field on a null mentor, and having to remember to handle the null in every case is rather annoying

granted, it probably is more semantically clear to have a section w/o an assigned teacher have a null mentor, so this is fine as is i guess